es/sources_index.go

```diff
@@ -0,0 +1,343 @@
+			if err != nil {
+				return []string{}, err
+			}
+			removed[source.MDB_UID] = true
+		}
+		delRes, err := index.esc.DeleteByQuery(indexName).
+			Query(elasticScope).
+			Do(context.TODO())
+		if err != nil {
+			return []string{}, errors.Wrapf(err, "Remove from index %s %+v\n", indexName, elasticScope)
+		}
+		if delRes.Deleted > 0 {
+			fmt.Printf("Deleted %d documents from %s.\n", delRes.Deleted, indexName)
+		}
+	}
+	if len(removed) == 0 {
+		fmt.Println("Sources Index - Nothing was delete.")
+		return []string{}, nil
+	}
+	keys := make([]string, 0)
+	for k := range removed {
+		keys = append(keys, k)
+	}
+	return keys, nil
+}
+
+func (index *SourcesIndex) getDocxPath(uid string, lang string) (string, error) {
+	uidPath := path.Join(sourcesFolder, uid)
+	jsonPath := path.Join(uidPath, "index.json")
+	jsonCnt, err := ioutil.ReadFile(jsonPath)
+	if err != nil {
+		return "", fmt.Errorf("Unable to read from file %s. Error: %+v", jsonPath, err)
+	}
+	var m map[string]map[string]string
+	err = json.Unmarshal(jsonCnt, &m)
+	if err != nil {
+		return "", err
+	}
+	if val, ok := m[lang]; ok {
+		docxPath := path.Join(uidPath, val["docx"])
+		if _, err := os.Stat(docxPath); err == nil {
+			return path.Join(docxPath), nil
+		}
+	}
+	return "", errors.New("Docx not found in index.json")
+}
+
+func (index *SourcesIndex) indexSource(mdbSource *mdbmodels.Source, parentsMap []string) error {
+	// Create documents in each language with available translation
+	hasDocxForSomeLanguage := false
+	i18nMap := make(map[string]Source)
+	for _, i18n := range mdbSource.R.SourceI18ns {
+
+		if i18n.Name.Valid && i18n.Name.String != "" {
+
+			source := Source{
+				MDB_UID: mdbSource.UID,
+				Name:    i18n.Name.String,
+				Sources: parentsMap,
+			}
+
+			if i18n.Description.Valid && i18n.Description.String != "" {
+				source.Description = i18n.Description.String
+			}
+
+			fPath, err := index.getDocxPath(mdbSource.UID, i18n.Language)
+			if err != nil {
+				log.Warnf("Unable to retrieving docx path for source %s with language %s. Skipping indexing.", mdbSource.UID, i18n.Language)
+				continue
+			} else {
+				content, err := index.ParseDocx(fPath)
+				if err != nil {
+					log.Warnf("Error parsing docx for source %s and language %s. Skipping indexing.", mdbSource.UID, i18n.Language)
+					continue
+				}
+				source.Content = content
+				hasDocxForSomeLanguage = true
+			}
+
+			for _, a := range mdbSource.R.Authors {
+
+				ai18n, err := mdbmodels.FindAuthorI18n(index.db, a.ID, i18n.Language)
+				if err != nil {
+					if err == sql.ErrNoRows {
+						continue
+					}
+					return err
+				}
+				if ai18n.Name.Valid && ai18n.Name.String != "" {
+					source.Authors = append(source.Authors, ai18n.Name.String)
+				}
+				if ai18n.FullName.Valid && ai18n.FullName.String != "" {
+					source.Authors = append(source.Authors, ai18n.FullName.String)
+				}
+			}
+
+			i18nMap[i18n.Language] = source
+		}
+	}
+
+	if !hasDocxForSomeLanguage {
```

Comment:
This should not stop the index process.
Some sources in the MDB don't have any docx at all.

Reply:
Done.